tools/communication/src/spi.cpp

```diff
@@ -0,0 +1,288 @@
+
+		m2KSpiDesc->clock = m2KSpiInit->clock;
+		m2KSpiDesc->mosi = m2KSpiInit->mosi;
+		m2KSpiDesc->miso = m2KSpiInit->miso;
+		m2KSpiDesc->bit_numbering = m2KSpiInit->bit_numbering;
+		m2KSpiDesc->context = m2KSpiInit->context;
+		m2KSpiDesc->digital = m2KSpiDesc->context->getDigital();
+		m2KSpiDesc->sample_rate = sampleRate;
+
+		m2KSpiDesc->digital->flushBufferIn();
+		m2KSpiDesc->digital->setKernelBuffersCountIn(1);
+
+		//set sampling frequencies
+		m2KSpiDesc->digital->setSampleRateOut(m2KSpiDesc->sample_rate);
+		m2KSpiDesc->digital->setSampleRateIn(m2KSpiDesc->sample_rate);
+
+		//enable the channels
+		setOutputChannel(spiDesc->chip_select, m2KSpiDesc->digital);
+		setOutputChannel(m2KSpiDesc->clock, m2KSpiDesc->digital);
+		setOutputChannel(m2KSpiDesc->mosi, m2KSpiDesc->digital);
+		setInputChannel(m2KSpiDesc->miso, m2KSpiDesc->digital);
+
+		m2KSpiDesc->digital->setOutputMode(spiDesc->chip_select, libm2k::digital::DIO_PUSHPULL);
+		m2KSpiDesc->digital->setOutputMode(m2KSpiDesc->clock, libm2k::digital::DIO_PUSHPULL);
+		m2KSpiDesc->digital->setOutputMode(m2KSpiDesc->mosi, libm2k::digital::DIO_PUSHPULL);
+		m2KSpiDesc->digital->setOutputMode(m2KSpiDesc->miso, libm2k::digital::DIO_PUSHPULL);
+
+		m2KSpiDesc->digital->setCyclic(false);
+
+		libm2k::analog::M2kHardwareTrigger *trigger = m2KSpiDesc->digital->getTrigger();
+		trigger->setDigitalCondition(m2KSpiDesc->clock, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(m2KSpiDesc->mosi, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(m2KSpiDesc->miso, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalCondition(spiDesc->chip_select, libm2k::analog::NO_TRIGGER);
+		trigger->setDigitalMode(libm2k::digital::DIO_OR);
+		trigger->setDigitalDelay(0);
+
+		//set the state of CS and CLK to idle
+		m2KSpiDesc->digital->setValueRaw(spiDesc->chip_select, libm2k::digital::HIGH);
+		if (spiDesc->mode >> 1u) {
+			m2KSpiDesc->digital->setValueRaw(m2KSpiDesc->clock, libm2k::digital::HIGH);
+		} else {
+			m2KSpiDesc->digital->setValueRaw(m2KSpiDesc->clock, libm2k::digital::LOW);
+		}
+
+		spiDesc->extra = (void *) m2KSpiDesc;
+		*desc = spiDesc;
+
+	} catch (std::exception &e) {
+		std::cout << e.what();
+		return -1;
+	}
+	return 0;
+}
+
+int32_t spi_remove(struct spi_desc *desc)
+{
+	try {
+		auto m2KSpiDesc = (m2k_spi_desc *) desc->extra;
+		delete m2KSpiDesc;
+		m2KSpiDesc = nullptr;
+		delete desc;
+		desc = nullptr;
+	} catch (std::exception &e) {
+		std::cout << e.what();
+		return -1;
+	}
+	return 0;
+}
+
+int32_t spi_write_and_read(struct spi_desc *desc,
+			   uint8_t *data,
+			   uint8_t bytes_number)
+{
+	try {
+		auto *m2KSpiDesc = (m2k_spi_desc *) desc->extra;
+		//start reading - wait until buffer is pushed
+		std::thread thread_read(read, desc, data, bytes_number);
+		std::vector<unsigned short> buffer = createBuffer(desc, data, bytes_number);
+		//make sure the reading thread is waiting
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
100ms seems like a long delay to arm the trigger & prepare the buffer.
@AlexandraTrifan mentioned errno from libiio could be leveraged here.

Reply:
The sleeping time was reduced to 1 ms.